File: core/src/main/java/com/mcreater/amclcore/concurrent/TaskState.java
package com.mcreater.amclcore.concurrent;

import com.mcreater.amclcore.i18n.Text;
import lombok.Builder;
import lombok.Data;

import java.util.function.Consumer;

@Data
@Builder
public class TaskState<R> {
    @Builder.Default
    private Type taskType = Type.EXECUTING;
    private Throwable throwable;
    private R result;
    @Builder.Default
    private int totalStage = 1;
    @Builder.Default
    private int currentStage = 1;
    private Text message;

    public double getStageDouble() {
        return totalStage != 0 ? (double) currentStage / totalStage : 0;
    }

    public int getStateInt() {
        return (int) (getStageDouble() * 100);
    }

    public enum Type {
        ERROR,
        EXECUTING,
        FINISHED
    }

    public boolean isExc() {
        return taskType == Type.ERROR;
    }

    public void executeIfExc(Consumer<Throwable> c) {
        if (isExc() && throwable != null) c.accept(throwable);
    }

    public boolean isFinish() {
        return taskType == Type.FINISHED;
    }

    public void executeIfFinish(Consumer<R> c) {
        if (isFinish() && result != null) c.accept(result);
    }
}

File: core/src/main/java/com/mcreater/amclcore/concurrent/TaskState.kt
package com.mcreater.amclcore.concurrent

import com.mcreater.amclcore.i18n.Text
import java.util.function.Consumer

data class TaskState<R>(
    var taskType: Type = Type.EXECUTING,
    var throwable: Throwable? = null,
    var result: R? = null,
    var totalStage: Int = 1,
    var currentStage: Int = 1,
    var message: Text? = null
) {
    enum class Type {
        ERROR,
        EXECUTING,
        FINISHED
    }

    fun getStageDouble(): Double {
        return if (totalStage != 0) currentStage.toDouble() / totalStage else 0.0
    }

    fun getStateInt(): Int {
        return (getStageDouble() * 100).toInt()
    }

    fun isExc(): Boolean {
        return taskType == Type.ERROR
    }

    fun executeIfExc(c: Consumer<Throwable?>) {
        if (isExc() && throwable != null) c.accept(throwable)
    }

    fun isFinish(): Boolean {
        return taskType == Type.FINISHED
    }

    fun executeIfFinish(c: Consumer<R>) {
        if (isFinish() && result != null) c.accept(result!!)
    }
}
